4.4 - Improve error logging

markstory opened the pull request:
Fix several error logging related issues that @ADmad identified.

* The `log` option now works in `ErrorTrap`
* Improved documentation for what `stderr` does.
* The `trace` option now works with `ErrorTrap`.
* Removed a redundant copy of the `skipLog` logic.
* Removed the configuration mutation when handling fatal errors. Instead the `trace` option is applied consistently.

markstory committed:
Fix log config option in ErrorTrap not working.

Also fix a few tests that weren't really doing what they should.

markstory committed:
Implement the `log` option for ExceptionTrap.

Remove the config mutation as it doesn't actually do anything. The fatal
error that is triggered is caught by the global error handler not the
exception handler.

markstory committed:
Consolidate skipLog code and improve documentation.

We don't need this copy of skipLog as ErrorLogger handles this logic as
well. Having it in the logger allows userland code to replace the
behavior more easily if they want to.

markstory committed:
Fix missing trace option support in ErrorTrap.

markstory committed:
Remove doubled trace data.

If traces are needed on fatal errors the `trace` option will enable the
collection of traces.

othercorey commented:
So now we want to only forward skipLog? Isn't that essentially the same as the revert PR I created earlier?

ADmad commented on src/Error/ErrorTrap.php:
```diff
@@ -123,15 +124,26 @@ public function handleError(
         $renderer = $this->renderer();
         $logger = $this->logger();
 
+        $context = [];
```
Logging context is used in the `try` block below only if logging is enabled, so this can be moved below that check.

ADmad commented on src/Error/ExceptionTrap.php:
```diff
@@ -36,20 +36,24 @@ class ExceptionTrap
      *   classes used in CLI and web environments you'll need to write that conditional logic as well.
      *   The conventional location for custom renderers is in `src/Error`. Your exception renderer needs to
      *   implement the `render()` method and return either a string or Http\Response.
+     * - `log` Set to false to disable logging.
```
Could you please similarly document all the options for `ErrorTrap::$_defaultConfig` too?

ADmad commented:
- `ErrorTrap::$_defaultConfig` has keys `extraFatalErrorMemory` and `ignoredDeprecationPaths` which are unused.
- `ExceptionTrap` uses `extraFatalErrorMemory` option but it's not listed in `$_defaultConfig`.

markstory commented:
> So now we want to only forward skipLog? Isn't that essentially the same as the revert PR I created earlier?

We could keep the other copy instead :shrug: I noticed last night that we had two copies of the code and didn't see a reason to have both. Do you know of a reason we should keep both?

markstory committed:
Fix mistakes in default configuration arrays.

markstory committed:
Only create context when we need it.

markstory committed:
Move skipLog handling to ExceptionTrap

This allows us to provide consistent log skipping for userland error
loggers as well.

markstory committed:
Improve docs for ErrorTrap

markstory committed:
Fix phpcs error.

othercorey commented:
> We could keep the other copy instead 🤷 I noticed last night that we had two copies of the code and didn't see a reason to have both. Do you know of a reason we should keep both?

I added the code to ExceptionTrap since I thought that would be a more appropriate place for skipLog and we could migrate to it instead of users having to configure individual loggers. 

However, it seemed that had side-effects so I was going to revert it. You then proposed re-writing it to filter out the logging but still call logException(). Now, we're just removing it entirely. 

I wasn't sure if there was a new motivation for all of this or of we're just going in circles with changes.

markstory commented:
> I wasn't sure if there was a new motivation for all of this or of we're just going in circles with changes.

We might be going in circles a bit. I think having the skipLog applied in ExceptionTrap is better as it won't need to be re-implemented in userland loggers. Sorry for all the back and forth around this, sometimes it takes me a few tries to get things sorted out in my head.

ADmad commented:
As I see it we have 3 decisions to make with regards to exception logging:
1. Should exceptions be logged?
2. Should logging this particular exception be skipped?
3. Should the logged message contain a trace?

If the `skipLog` check is going to be done in `ExceptionTrap` then it would be better to keep the `trace` check there too. `ErrorLogger` would then not need to hold any config of it's own and all required info would be passed as method arguments.

On a related note I would prefer if `ErrorLoggerInterface::log()` was named `ErrorLoggerInterface::logException()`.

markstory commented:
> As I see it we have 3 decisions to make with regards to exception logging

I agree with this list of decisions and that they can all happen in the Exception/Error trap classes. I can update this pull request to accommodate that.

> On a related note I would prefer if ErrorLoggerInterface::log() was named ErrorLoggerInterface::logException().

We should be able to do that in a backwards compatible way by deprecating log() and adding logException().

markstory committed:
Fix missing skipLog block.

We need this in both `BaseErrorHandler` and `ExceptionTrap` if we're
moving it out of the logger. Also fix a phpcs error.

markstory committed:
Further improve the ErrorLoggerInterface

Provide better method names, and better parameters. This should also
help reduce the need for error handling configuration to be used in
Loggers which feels like a win as userland code doesn't need to read
configuration data to make decisions on whether or not to include
a trace.

The new methods can't be added to the interface because of backwards
compatibility so I've annotated them and used `method_exists()`. This
isn't ideal but it does allow us to be backwards compatible.

markstory commented:
I did some more invasive changes in 165bb75. I've added new methods with 'cleaner' signatures and passed in the new error object, normalized the method names and extracted the logging, skiplog and trace inclusion decisions.

These changes while 'cleaner' do make upgrading harder as the previous interface methods are 'deprecated' and emitting errors. If folks feel we shouldn't include deprecation warnings I'm up for removing them as it is disruptive change and could make adopting the new error subsystem harder.

ADmad commented:
I like these changes. Only nitpick I have is perhaps the `$includeTrace` and `$request` argument for the 2 new error logger methods should be swapped?

> If folks feel we shouldn't include deprecation warnings I'm up for removing them as it is disruptive change and could make adopting the new error subsystem harder.

Not sure how often people need to customize/override error/exception handling. Personally I have only needed to customize the exception renderer.

ADmad commented on src/Error/ErrorLogger.php:
```diff
@@ -55,7 +52,62 @@ public function __construct(array $config = [])
+     * @param \Cake\Error\PhpError $error The error to log
+     * @param bool $includeTrace Should the log message include a stacktrace
+     * @param ?\Psr\Http\Message\ServerRequestInterface $request The request if in an HTTP context.
+     * @return bool Logging success
```
Does the bool return have any practical use? The callers simply ignore the return anyway.

markstory replied:
That's fair. We can do void

markstory commented:
> Not sure how often people need to customize/override error/exception handling. Personally I have only needed to customize the exception renderer.

Swapping out the error logger is useful when integrating an error monitoring system :) But outside of that I don't see it as something users would need to do.

markstory committed:
Swap order of parameters in error logger methods and remove return

We don't have any callers that use the return value of these methods and
until we do there isn't really a reason to have a return value.

othercorey commented:
> > Not sure how often people need to customize/override error/exception handling. Personally I have only needed to customize the exception renderer.
> 
> Swapping out the error logger is useful when integrating an error monitoring system :) But outside of that I don't see it as something users would need to do.

I have only needed to push exceptions to an external service for logging.

markstory commented:
One thought I had today was that instead of using annotated methods and `method_exists` checks we could add another interface `ErrorLoggingInterface` (not :100: on the name). With a separate interface we can use `instanceof` checks.

The core provided logger could implement both interfaces, and user-land implementations could do the same. Do folks think that would be a 'cleaner' approach over the method annotations and `method_exists()` checks? I don't think it would change the shape of the code in `ErrorTrap` and `ExceptionTrap`, or the deprecations we emit, so perhaps it is over-engineering.

markstory committed:
Add more deprecation notices.

markstory committed:
Update baseline files.

ADmad commented:
> we could add another interface ErrorLoggingInterface

Given that we are on probably the last minor release for 4.x I don't think a new interface is necessary.
